File: backend/api/strategies/state.py
from __future__ import annotations

from dataclasses import dataclass, field

from api.domain.curation import JournalEntry
from api.domain.graph import Edge, Node


@dataclass
class GraphBuildState:
    """In-memory graph as it flows through builder → cleaners → clusterer.

    Domain entities (Node, Edge) keep their identity across stages so
    provenance survives. The journal accumulates structural ops applied
    along the way; once persisted (Phase 2), it becomes the variant's
    initial JournalEntry stream.

    Cleaners are pure-ish: they consume one state and produce another.
    No DB I/O happens until the orchestrator persists at the end of the
    pipeline.
    """

    nodes: list[Node] = field(default_factory=list)
    edges: list[Edge] = field(default_factory=list)
    journal: list[JournalEntry] = field(default_factory=list)

    def node_index(self) -> dict:
        return {n.id: n for n in self.nodes}

    def edge_index(self) -> dict:
        return {e.id: e for e in self.edges}

File: packages/graph-core/src/engine/state.ts
import type { Node, Edge } from "../domain/graph.js";
import type { JournalEntry } from "../domain/journal.js";
import type { Id } from "../domain/types.js";

/** The graph as it flows through curation: nodes + edges + the journal of
 *  ops applied so far. Pure value object — appliers return a new state. */
export interface GraphState {
  nodes: Node[];
  edges: Edge[];
  journal: JournalEntry[];
}

export const emptyState = (): GraphState => ({
  nodes: [],
  edges: [],
  journal: [],
});

export const nodeIndex = (s: GraphState): Map<Id, Node> =>
  new Map(s.nodes.map((n) => [n.id, n]));

export const edgeIndex = (s: GraphState): Map<Id, Edge> =>
  new Map(s.edges.map((e) => [e.id, e]));

/** Community id a node belongs to via its MEMBER_OF edge (or null). */
export const communityOf = (s: GraphState, nodeId: Id): Id | null => {
  for (const e of s.edges) {
    if (e.type === "member_of" && e.sourceId === nodeId) return e.targetId;
  }
  return null;
};
